---
title: "R Notebook"
output: html_notebook
---

```{r}
### SNPs with genotype decided by RNA-reads
freq <-read.table("SNP_geno_freq.txt", sep = "\t", header = T)
freq['exp_homo'] <- 0.5
freq['exp_hete'] <- 0.5
freq.num <- freq[, c("homo", "hete", "exp_homo", "exp_hete")]
row.names(freq.num) <- with(freq, paste(chromosome, position, sep = ":"))
```

```{r}
### 
get_chisq <- function(x, probs) {
  chsq <- chisq.test(x = x, p = probs)
  ans <- cbind(statistic = chsq$statistic[[1]], p.value = chsq$p.value)
  ans
}
sol <- data.frame(t(apply(freq.num, 1, function(x) get_chisq(x[1:2], x[3:4]))))
names(sol) <- c("statistic", "pvalue")
###
freq_p <- cbind(freq.num, sol)
freq_p['padj'] <- p.adjust(freq_p$pvalue, method = "BH")
freq_p <- freq_p[, c("homo", "hete", "pvalue", "padj")]
write.table(freq_p, file = "../data/geno_p.txt", sep = "\t", quote = F, row.names = T)
### SNPs with significance deviated from theoretical ratio 
sig_p <- freq_p[freq_p$padj < 0.05, ] # padj 0.05
write.table(sig_p, file = "../data/sig_p0.05.txt", sep = "\t", quote = F, row.names = T)
sig_p <- freq_p[freq_p$padj < 0.01, ] # padj 0.01
write.table(sig_p, file = "../data/sig_p0.01.txt", sep = "\t", quote = F, row.names = T)
```

```{r}
### SNPs with RNA-reads supporting and meet 1:1 ratio
RNA_p <- read.table("geno_p.txt", sep = "\t", header = T)
RNA_quality <- RNA_p[RNA_p$homo > 5 & RNA_p$hete > 5 & RNA_p$padj > 1e-2, ]
write.table(RNA_quality, file = "RNA_SNP_quality.txt", sep = "\t", quote = F, row.names = T)
```

